### ISSUE TITLE
file.managed fails for long URLs

### ISSUE THREAD

**Issue by ghost** [saltstack/salt#40846]
### Description of Issue/Question

`file.managed` fails when using long URLs, such as signed URLs with Google Cloud:

    https://storage.googleapis.com/example-bucket/cat.jpeg?GoogleAccessId=example@example-project.iam.gserviceaccount.com&Expires=1458238630&Signature=VVUgfqviDCov%2B%2BKnmVOkwBR2olSbId51kSibuQeiH8ucGFyOfAVbH5J%2B5V0gDYIioO2dDGH9Fsj6YdwxWv65HE71VEOEsVPuS8CVb%2BVeeIzmEe8z7X7o1d%2BcWbPEo4exILQbj3ROM3T2OrkNBU9sbHq0mLbDMhiiQZ3xCaiCQdsrMEdYVvAFggPuPq%2FEQyQZmyJK3ty%2Bmr7kAFW16I9pD11jfBSD1XXjKTJzgd%2FMGSde4Va4J1RtHoX7r5i7YR7Mvf%2Fb17zlAuGlzVUf%2FzmhLPqtfKinVrcqdlmamMcmLoW8eLG%2B1yYW%2F7tlS2hvqSfCW8eMUUjiHiSWgZLEVIG4Lw%3D%3D

with this state:

    download_cat_jpeg:
    file.managed:
      - name: /tmp/cat.jpeg
      - source: https://storage.googleapis.com/example-bucket/cat.jpeg?GoogleAccessId=example@example-project.iam.gserviceaccount.com&Expires=1458238630&Signature=VVUgfqviDCov%2B%2BKnmVOkwBR2olSbId51kSibuQeiH8ucGFyOfAVbH5J%2B5V0gDYIioO2dDGH9Fsj6YdwxWv65HE71VEOEsVPuS8CVb%2BVeeIzmEe8z7X7o1d%2BcWbPEo4exILQbj3ROM3T2OrkNBU9sbHq0mLbDMhiiQZ3xCaiCQdsrMEdYVvAFggPuPq%2FEQyQZmyJK3ty%2Bmr7kAFW16I9pD11jfBSD1XXjKTJzgd%2FMGSde4Va4J1RtHoX7r5i7YR7Mvf%2Fb17zlAuGlzVUf%2FzmhLPqtfKinVrcqdlmamMcmLoW8eLG%2B1yYW%2F7tlS2hvqSfCW8eMUUjiHiSWgZLEVIG4Lw%3D%3D
      - source_hash: ef07d5ad07267e155f418ff41696b9ba20887ae9e04c9a34c294ce5cc0e7872a

with the error:

     Comment: Failed to cache https://storage.googleapis.com/example-bucket/cat.jpeg?GoogleAccessId=example@example-project.iam.gserviceaccount.com&Expires=1458238630&Signature=VVUgfqviDCov%2B%2BKnmVOkwBR2olSbId51kSibuQeiH8ucGFyOfAVbH5J%2B5V0gDYIioO2dDGH9Fsj6YdwxWv65HE71VEOEsVPuS8CVb%2BVeeIzmEe8z7X7o1d%2BcWbPEo4exILQbj3ROM3T2OrkNBU9sbHq0mLbDMhiiQZ3xCaiCQdsrMEdYVvAFggPuPq%2FEQyQZmyJK3ty%2Bmr7kAFW16I9pD11jfBSD1XXjKTJzgd%2FMGSde4Va4J1RtHoX7r5i7YR7Mvf%2Fb17zlAuGlzVUf%2FzmhLPqtfKinVrcqdlmamMcmLoW8eLG%2B1yYW%2F7tlS2hvqSfCW8eMUUjiHiSWgZLEVIG4Lw%3D%3D: [Errno 36] File name too long: '/var/cache/salt/minion/extrn_files/base/storage.googleapis.com/example-bucket/cat.jpeg?GoogleAccessId=example@example-project.iam.gserviceaccount.com&Expires=1458238630&Signature=VVUgfqviDCov%2B%2BKnmVOkwBR2olSbId51kSibuQeiH8ucGFyOfAVbH5J%2B5V0gDYIioO2dDGH9Fsj6YdwxWv65HE71VEOEsVPuS8CVb%2BVeeIzmEe8z7X7o1d%2BcWbPEo4exILQbj3ROM3T2OrkNBU9sbHq0mLbDMhiiQZ3xCaiCQdsrMEdYVvAFggPuPq%2FEQyQZmyJK3ty%2Bmr7kAFW16I9pD11jfBSD1XXjKTJzgd%2FMGSde4Va4J1RtHoX7r5i7YR7Mvf%2Fb17zlAuGlzVUf%2FzmhLPqtfKinVrcqdlmamMcmLoW8eLG%2B1yYW%2F7tlS2hvqSfCW8eMUUjiHiSWgZLEVIG4Lw%3D%3D.part'

I think this can be solved with a sha256 hash for the filename of the part file, or a sha256 hash of the URL.

**Comment by Ch3LL** [comment]
I'm able to replicate this. And as i believe you hinted and expected we are getting this because the filename is too long 

```
File "/testing/salt/utils/__init__.py", line 1307, in fopen
    fhandle = open(*args, **kwargs)
IOError: [Errno 36] File name too long: '/var/cache/salt/minion/extrn_files/base/www.longur*
```

Seems maybe we could add an argument to the module that caches the hash and not the name of the file as you stated. Or an argument that allows you to supply a different name to store the file on the host

**Comment by Ch3LL** [comment]
Also here is a docker container to easily replicate the issue.

1. `docker run -ti -v /home/ch3ll/git/salt/:/testing/ ch3ll/issues:40846 salt-call --local state.sls test -ldebug` (where /home/ch3ll/git/salt is a local cloned git repo of salt)

**Comment by terminalmage** [comment]
Fix was merged in #45808 and backported in #47673.
